File: go/retry.go
package lasersell

import (
	"context"
	"errors"
	"fmt"
	"time"
)

// RetryPolicy controls retry attempts and exponential backoff behavior.
type RetryPolicy struct {
	// Maximum number of attempts including the first attempt.
	MaxAttempts int
	// Delay used before the first retry.
	InitialBackoff time.Duration
	// Upper bound for exponential backoff delay growth.
	MaxBackoff time.Duration
	// Maximum random jitter added to each retry delay.
	Jitter time.Duration
}

// LowLatencyRetryPolicy returns defaults suitable for short-lived API requests.
func LowLatencyRetryPolicy() RetryPolicy {
	return RetryPolicy{
		MaxAttempts:    2,
		InitialBackoff: 25 * time.Millisecond,
		MaxBackoff:     100 * time.Millisecond,
		Jitter:         25 * time.Millisecond,
	}
}

// DelayForAttempt computes the delay for the given 1-based attempt index.
func (p RetryPolicy) DelayForAttempt(attempt int) time.Duration {
	normalized := p.normalized()
	delay := normalized.InitialBackoff
	for i := 1; i < attempt; i++ {
		delay *= 2
		if delay > normalized.MaxBackoff {
			delay = normalized.MaxBackoff
		}
	}
	return delay + jitterDuration(normalized.Jitter, attempt)
}

func (p RetryPolicy) normalized() RetryPolicy {
	out := p
	if out.MaxAttempts <= 0 {
		out.MaxAttempts = 1
	}
	if out.InitialBackoff < 0 {
		out.InitialBackoff = 0
	}
	if out.MaxBackoff < 0 {
		out.MaxBackoff = 0
	}
	if out.MaxBackoff > 0 && out.InitialBackoff > out.MaxBackoff {
		out.InitialBackoff = out.MaxBackoff
	}
	if out.Jitter < 0 {
		out.Jitter = 0
	}
	return out
}

// Retry executes op with retries controlled by policy.
//
// op receives the 1-based attempt index. shouldRetry decides whether an error
// should trigger another attempt.
func Retry[T any](
	ctx context.Context,
	policy RetryPolicy,
	op func(attempt int) (T, error),
	shouldRetry func(error) bool,
) (T, error) {
	var zero T
	normalized := policy.normalized()

	for attempt := 1; attempt <= normalized.MaxAttempts; attempt++ {
		value, err := op(attempt)
		if err == nil {
			return value, nil
		}

		if attempt >= normalized.MaxAttempts || !shouldRetry(err) {
			return zero, err
		}

		delay := normalized.DelayForAttempt(attempt)
		if delay <= 0 {
			continue
		}

		timer := time.NewTimer(delay)
		select {
		case <-ctx.Done():
			timer.Stop()
			return zero, ctx.Err()
		case <-timer.C:
		}
	}

	return zero, errors.New("retry reached an unreachable state")
}

// TimeoutError indicates an operation exceeded its deadline.
type TimeoutError struct {
	Timeout time.Duration
}

func (e TimeoutError) Error() string {
	return fmt.Sprintf("operation timed out after %s", e.Timeout)
}

func (e TimeoutError) Is(target error) bool {
	_, ok := target.(TimeoutError)
	return ok
}

// WithTimeout applies timeout to fn while preserving parent cancellation.
func WithTimeout[T any](
	ctx context.Context,
	timeout time.Duration,
	fn func(ctx context.Context) (T, error),
) (T, error) {
	if timeout <= 0 {
		return fn(ctx)
	}

	timeoutCtx, cancel := context.WithTimeout(ctx, timeout)
	defer cancel()

	value, err := fn(timeoutCtx)
	if err != nil {
		if errors.Is(timeoutCtx.Err(), context.DeadlineExceeded) {
			var zero T
			return zero, TimeoutError{Timeout: timeout}
		}
		return value, err
	}

	return value, nil
}

func jitterDuration(maxJitter time.Duration, attempt int) time.Duration {
	if maxJitter <= 0 {
		return 0
	}

	limitNanos := uint64(maxJitter.Nanoseconds())
	if limitNanos == 0 {
		return 0
	}

	nowNanos := uint64(time.Now().UnixNano())
	mixed := nowNanos ^ (uint64(attempt) * 0x9e3779b97f4a7c15)
	return time.Duration(mixed % (limitNanos + 1))
}

File: rust/src/retry.rs
//! Retry and timeout utilities.
//!
//! The helpers in this module are transport-agnostic and are used by clients
//! that need bounded retries with lightweight jitter.

use std::future::Future;
use std::time::{Duration, SystemTime, UNIX_EPOCH};

use tracing::debug;

/// Policy controlling retry attempts and exponential backoff behavior.
#[derive(Clone, Debug)]
pub struct RetryPolicy {
    /// Maximum number of attempts including the first attempt.
    pub max_attempts: usize,
    /// Delay used before the first retry.
    pub initial_backoff: Duration,
    /// Upper bound for exponential backoff delay growth.
    pub max_backoff: Duration,
    /// Maximum random jitter added to each retry delay.
    pub jitter: Duration,
}

impl RetryPolicy {
    /// Returns a low-latency default suitable for short-lived API requests.
    pub fn low_latency() -> Self {
        Self {
            max_attempts: 2,
            initial_backoff: Duration::from_millis(25),
            max_backoff: Duration::from_millis(100),
            jitter: Duration::from_millis(25),
        }
    }

    /// Computes the delay to apply before the given retry attempt.
    ///
    /// `attempt` is 1-based and should correspond to the current attempt index.
    pub fn delay_for_attempt(&self, attempt: usize) -> Duration {
        let mut delay = self.initial_backoff;
        for _ in 1..attempt {
            delay = std::cmp::min(delay.saturating_mul(2), self.max_backoff);
        }
        delay + jitter_duration(self.jitter, attempt)
    }
}

impl Default for RetryPolicy {
    fn default() -> Self {
        Self::low_latency()
    }
}

/// Executes an async operation with retry behavior controlled by `policy`.
///
/// `op` receives the 1-based attempt number and must return a future that
/// resolves to the operation result. `should_retry` determines whether each
/// error is retryable.
pub async fn retry_async<T, E, Op, Fut, ShouldRetry>(
    policy: &RetryPolicy,
    mut op: Op,
    mut should_retry: ShouldRetry,
) -> Result<T, E>
where
    Op: FnMut(usize) -> Fut,
    Fut: Future<Output = Result<T, E>>,
    ShouldRetry: FnMut(&E) -> bool,
{
    let max_attempts = policy.max_attempts.max(1);

    for attempt in 1..=max_attempts {
        match op(attempt).await {
            Ok(value) => return Ok(value),
            Err(error) => {
                if attempt >= max_attempts || !should_retry(&error) {
                    return Err(error);
                }

                let delay = policy.delay_for_attempt(attempt);
                debug!(
                    event = "retry_attempt_failed",
                    attempt,
                    max_attempts,
                    delay_ms = delay.as_millis() as u64
                );
                if !delay.is_zero() {
                    tokio::time::sleep(delay).await;
                }
            }
        }
    }

    unreachable!("max_attempts is always at least 1")
}

/// Applies a timeout to an async computation.
pub async fn with_timeout<T, Fut>(
    timeout: Duration,
    future: Fut,
) -> Result<T, tokio::time::error::Elapsed>
where
    Fut: Future<Output = T>,
{
    tokio::time::timeout(timeout, future).await
}

fn jitter_duration(max_jitter: Duration, attempt: usize) -> Duration {
    if max_jitter.is_zero() {
        return Duration::ZERO;
    }

    let limit_nanos = max_jitter.as_nanos().min(u64::MAX as u128) as u64;
    if limit_nanos == 0 {
        return Duration::ZERO;
    }

    let now_nanos = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default()
        .subsec_nanos() as u64;
    let mixed = now_nanos ^ ((attempt as u64).wrapping_mul(0x9E37_79B9_7F4A_7C15));
    Duration::from_nanos(mixed % (limit_nanos + 1))
}

#[cfg(test)]
mod tests {
    use std::sync::atomic::{AtomicUsize, Ordering};
    use std::sync::Arc;
    use std::time::Duration;

    use super::{retry_async, RetryPolicy};

    #[test]
    fn retries_until_success() {
        let runtime = tokio::runtime::Builder::new_current_thread()
            .enable_time()
            .build()
            .expect("runtime");

        runtime.block_on(async {
            let calls = Arc::new(AtomicUsize::new(0));
            let policy = RetryPolicy {
                max_attempts: 3,
                initial_backoff: Duration::from_millis(1),
                max_backoff: Duration::from_millis(2),
                jitter: Duration::from_millis(0),
            };

            let result = retry_async(
                &policy,
                {
                    let calls = Arc::clone(&calls);
                    move |_| {
                        let calls = Arc::clone(&calls);
                        async move {
                            let value = calls.fetch_add(1, Ordering::SeqCst);
                            if value < 2 {
                                Err("retry")
                            } else {
                                Ok("ok")
                            }
                        }
                    }
                },
                |_| true,
            )
            .await;

            assert_eq!(result.expect("success"), "ok");
            assert_eq!(calls.load(Ordering::SeqCst), 3);
        });
    }

    #[test]
    fn stops_when_retry_predicate_rejects() {
        let runtime = tokio::runtime::Builder::new_current_thread()
            .enable_time()
            .build()
            .expect("runtime");

        runtime.block_on(async {
            let calls = Arc::new(AtomicUsize::new(0));
            let policy = RetryPolicy {
                max_attempts: 5,
                initial_backoff: Duration::from_millis(1),
                max_backoff: Duration::from_millis(2),
                jitter: Duration::from_millis(0),
            };

            let result: Result<(), &str> = retry_async(
                &policy,
                {
                    let calls = Arc::clone(&calls);
                    move |_| {
                        let calls = Arc::clone(&calls);
                        async move {
                            calls.fetch_add(1, Ordering::SeqCst);
                            Err("fatal")
                        }
                    }
                },
                |_| false,
            )
            .await;

            assert_eq!(result.expect_err("expected failure"), "fatal");
            assert_eq!(calls.load(Ordering::SeqCst), 1);
        });
    }
}
